src/cosim/utility/thread_pool.hpp

```diff
@@ -0,0 +1,88 @@
+
+#ifndef COSIM_UTILITY_THREAD_POOL_HPP
+#define COSIM_UTILITY_THREAD_POOL_HPP
+
+#include <atomic>
+#include <condition_variable>
+#include <functional>
+#include <mutex>
+#include <queue>
+#include <thread>
+
+namespace cosim
+{
+namespace utility
+{
+
+class thread_pool
+{
+private:
+    std::atomic_bool done_;
+    std::queue<std::function<void()>> work_queue_;
+    std::vector<std::thread> threads_;
+    std::mutex m_;
+    std::condition_variable cv_;
+
+    void worker_thread()
+    {
+        while (!done_) {
+
+            std::unique_lock<std::mutex> lck(m_);
+            if (!work_queue_.empty()) {
+
+                auto task = work_queue_.front();
+                work_queue_.pop();
+                task();
+
+                lck.unlock();
+                cv_.notify_one();
+
+            } else {
+                std::this_thread::yield();
+            }
```

Comment:
Thanks for working on this PR. Would it be possible to have an option to run fmus without spawning a separate worker thread? (i.e. running everything in a main thread sequentially). For some use-cases, the communication overhead between the main and the worker threads seem quite significant.

Reply:
This is possible today, and was in fact the motivation for using fibers in the first place. (Not that fibers are required for sequential execution of FMU functions, but they provide a nice framework for combining sequential execution with asynchronous I/O so one can run both local and remote slaves without the overhead of threads.)

Reply:
The problem is just that the remote-slave implementation we're using today (proxyfmu) is not built around async I/O, so it requires a worker thread per slave anyway. That is why fibers simply are an extra overhead that comes _in addition to_ thread and I/O today, which is what I suppose motivated @markaren's PR.

As I see it, there are two ways to improve performance in this area: Drop fibers and go all-in for proxyfmu and one-thread-per-slave (as this PR proposes), or keep fibers and implement async communication with slaves (as was the original plan).

Reply:
Thanks for the explanation @kyllingstad.

I was able to achieve required simulation performance by removing `std::thread` for slaves as well as recursive fiber creation in the libcosim master branch. I was not able to achieve this via `pseudo_async_slave` because it still creates fiber recursively in the fmu interface methods (calling get/setters for variables and do_step).

But now the problem is I cannot not simulate some fmus because they run directly in fibers with limited resources (created by `slave_simulator::impl::do_step`). And I cannot replace/extend `slave_simulator`  as it is always added in `execution::add_slave`.

Reply:
> Would it be possible to have an option to run fmus without spawning a separate worker thread?
Yes, and this is one of the motivations for this PR. IMO the master algorithm itself should decide how it handles execution. In the case of this implementation it could take `numThreads` as an argument where < 1 means no pool.

proxy-fmu was not a motivation for this PR. My motivation was to simplify the code base and make it run faster. I've found that the fiber solution is slower than a threaded solution with or without proxy-fmu.